pkg/object/udpproxy/runtime.go

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package udpproxy
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/supervisor"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/ipfilter"
+	"github.com/megaease/easegress/pkg/util/layer4backend"
+)
+
+type (
+	runtime struct {
+		superSpec *supervisor.Spec
+		spec      *Spec
+
+		pool       *layer4backend.Pool // backend servers pool
+		serverConn *net.UDPConn        // listener
+		sessions   map[string]*session
+
+		ipFilters *ipfilter.Layer4IpFilters
+
+		mu   sync.Mutex
+		done chan struct{}
+	}
+)
+
+func newRuntime(superSpec *supervisor.Spec) *runtime {
+	spec := superSpec.ObjectSpec().(*Spec)
+	r := &runtime{
+		superSpec: superSpec,
+		spec:      spec,
+
+		pool:      layer4backend.NewPool(superSpec.Super(), spec.Pool, ""),
+		ipFilters: ipfilter.NewLayer4IPFilters(spec.IPFilter),
+
+		sessions: make(map[string]*session),
+	}
+
+	r.startServer()
+	return r
+}
+
+// Close notify runtime close
+func (r *runtime) Close() {
+
+	close(r.done)
+	_ = r.serverConn.Close()
+
+	r.mu.Lock()
+	for k, s := range r.sessions {
+		delete(r.sessions, k)
+		s.Close()
+	}
+	r.sessions = nil
+	r.mu.Unlock()
+
+	r.pool.Close()
```

Comment:
we should do these cleanup operations until we are sure all other goroutines have stopped, using a lock here won't make the code goroutine safe in most cases.